packages/langchain/src/openai/util.ts

```diff
@@ -0,0 +1,212 @@
+import {
+  AIMessage,
+  BaseMessage,
+  ChatMessage,
+  ToolMessage
+} from '@langchain/core/messages';
+import { ChatResult } from '@langchain/core/outputs';
+import { StructuredTool } from '@langchain/core/tools';
+import type {
+  OpenAiChatCompletionChoice,
+  OpenAiChatCompletionFunction,
+  OpenAiChatCompletionTool,
+  OpenAiChatMessage,
+  OpenAiChatCompletionOutput,
+  OpenAiChatCompletionParameters
+} from '@sap-ai-sdk/foundation-models';
+import { zodToJsonSchema } from 'zod-to-json-schema';
+import { OpenAiChatClient } from './chat.js';
+import { OpenAiChatCallOptions } from './types.js';
+
+/**
+ * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionFunction}.
+ * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
+ * @returns The OpenAI Chat Completion Function.
+ * @internal
+ */
+export function mapToolToOpenAiFunction(
+  tool: StructuredTool
+): OpenAiChatCompletionFunction {
+  return {
+    name: tool.name,
+    description: tool.description,
+    parameters: zodToJsonSchema(tool.schema)
+  };
+}
+
+/**
+ * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionTool}.
+ * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
+ * @returns The OpenAI Chat Completion Tool.
+ * @internal
+ */
+export function mapToolToOpenAiTool(
+  tool: StructuredTool
+): OpenAiChatCompletionTool {
+  return {
+    type: 'function',
+    function: {
+      name: tool.name,
+      description: tool.description,
+      parameters: zodToJsonSchema(tool.schema)
+    }
+  };
+}
+
+/**
+ * Maps a {@link BaseMessage} to OpenAI's Message Role.
+ * @param message - The message to map.
+ * @returns The OpenAI Message Role.
+ * @internal
+ */
+export function mapBaseMessageToRole(
```

Comment:
[q] isn't there some function inside `@langchain/openai` that we can use for this, and only re-export it with the correct type?

I.e. doing
```suggestion
export function mapToolToOpenAiTool( msg: BaseMessage): OpenAiChatMessage['role'] { 
  return langchainFunction(msg) as OpenAiChatMessage['role']; 
}
```